flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateHandle.java

```diff
@@ -0,0 +1,209 @@
+		this.keyGroupRange = keyGroupRange;
+		this.newSstFileNames = newSstFileNames;
+		this.sstFiles = sstFiles;
+		this.miscFiles = miscFiles;
+		this.metaStateHandle = metaStateHandle;
+		this.registered = false;
+	}
+
+	@Override
+	public KeyGroupRange getKeyGroupRange() {
+		return keyGroupRange;
+	}
+
+	public Map<String, StreamStateHandle> getSstFiles() {
+		return sstFiles;
+	}
+
+	public Map<String, StreamStateHandle> getMiscFiles() {
+		return miscFiles;
+	}
+
+	public StreamStateHandle getMetaStateHandle() {
+		return metaStateHandle;
+	}
+
+	@Override
+	public KeyedStateHandle getIntersection(KeyGroupRange keyGroupRange) {
+		if (this.keyGroupRange.getIntersection(keyGroupRange) != KeyGroupRange.EMPTY_KEY_GROUP_RANGE) {
+			return this;
+		} else {
+			return null;
+		}
+	}
+
+	@Override
+	public void discardState() throws Exception {
+
+		try {
+			metaStateHandle.discardState();
+		} catch (Exception e) {
+			LOG.warn("Could not properly discard meta data.", e);
+		}
+
+		try {
+			StateUtil.bestEffortDiscardAllStateObjects(miscFiles.values());
+		} catch (Exception e) {
+			LOG.warn("Could not properly discard misc file state.", e);
+		}
+
+		if (!registered) {
+			for (String newSstFileName : newSstFileNames) {
+				StreamStateHandle handle = sstFiles.get(newSstFileName);
+				try {
+					handle.discardState();
+				} catch (Exception e) {
+					LOG.warn("Could not properly discard sst file state", e);
+				}
+			}
+		}
+	}
+
+	@Override
+	public long getStateSize() {
+		long size = StateUtil.getStateSize(metaStateHandle);
+
+		for (StreamStateHandle sstFileHandle : sstFiles.values()) {
+			size += sstFileHandle.getStateSize();
+		}
+
+		for (StreamStateHandle miscFileHandle : miscFiles.values()) {
+			size += miscFileHandle.getStateSize();
+		}
+
+		return size;
+	}
+
+	@Override
+	public void registerSharedStates(SharedStateRegistry stateRegistry) {
+		Preconditions.checkState(!registered, "The state handle has already registered its shared states.");
+
+		for (Map.Entry<String, StreamStateHandle> sstFileEntry : sstFiles.entrySet()) {
+			SstFileStateHandle stateHandle = new SstFileStateHandle(sstFileEntry.getKey(), sstFileEntry.getValue());
+
+			int referenceCount = stateRegistry.register(stateHandle);
+
+			if (newSstFileNames.contains(sstFileEntry.getKey())) {
+				Preconditions.checkState(referenceCount == 1);
+			} else {
+				Preconditions.checkState(referenceCount > 1);
+			}
+		}
+
+		registered = true;
+	}
+
+	@Override
+	public void unregisterSharedStates(SharedStateRegistry stateRegistry) {
+		Preconditions.checkState(registered, "The state handle has not registered its shared states yet.");
+
+		for (Map.Entry<String, StreamStateHandle> sstFileEntry : sstFiles.entrySet()) {
+			stateRegistry.unregister(new SstFileStateHandle(sstFileEntry.getKey(), sstFileEntry.getValue()));
+		}
+
+		registered = false;
+	}
+
+	private class SstFileStateHandle implements SharedStateHandle {
```

Comment:
The `StreamStateHandle` returned by the `CheckpointStateOutputStream` may also be typed`ByteArrayStreamHandle`. So we cannot simply use `SharedFileStateHandle` here. I think we do not change the code here and do the refactoring in the future. What do you think?